lib/internal/quic/core.js

```diff
@@ -0,0 +1,3341 @@
+      throw new ERR_QUICSESSION_DESTROYED('startHandshake');
+    if (this.#handshakeStarted)
+      return;
+    this.#handshakeStarted = true;
+    if (!this.ready) {
+      this.once('ready', () => this[kHandle].startHandshake());
+    } else {
+      this[kHandle].startHandshake();
+    }
+  }
+
+  get ephemeralKeyInfo() {
+    return this[kHandle] !== undefined ?
+      this[kHandle].getEphemeralKeyInfo() :
+      {};
+  }
+
+  #setSocketAfterBind = function(socket, callback) {
+    if (socket.destroyed) {
+      callback(new ERR_QUICSOCKET_DESTROYED('setSocket'));
+      return;
+    }
+
+    if (!this[kHandle].setSocket(socket[kHandle])) {
+      callback(new ERR_QUICCLIENTSESSION_FAILED_SETSOCKET());
+      return;
+    }
+
+    if (this.socket) {
+      this.socket[kRemoveSession](this);
+      this[kSetSocket](undefined);
+    }
+    socket[kAddSession](this);
+    this[kSetSocket](socket);
+
+    callback();
+  };
+
+  setSocket(socket, callback) {
+    if (!(socket instanceof QuicSocket))
+      throw new ERR_INVALID_ARG_TYPE('socket', 'QuicSocket', socket);
+
+    if (typeof callback !== 'function')
+      throw new ERR_INVALID_CALLBACK();
+
+    socket[kMaybeBind](() => this.#setSocketAfterBind(socket, callback));
+  }
+}
+
+function streamOnResume() {
+  if (!this.destroyed)
+    this[kHandle].readStart();
+}
+
+function streamOnPause() {
+  if (!this.destroyed /* && !this.pending */)
+    this[kHandle].readStop();
+}
+
+class QuicStream extends Duplex {
+  #closed = false;
+  #aborted = false;
+  #defaultEncoding = undefined;
+  #didRead = false;
+  #id = undefined;
+  #highWaterMark = undefined;
+  #push_id = undefined;
+  #resetCode = undefined;
+  #session = undefined;
+  #dataRateHistogram = undefined;
+  #dataSizeHistogram = undefined;
+  #dataAckHistogram = undefined;
+  #stats = undefined;
+
+  constructor(options, session, push_id) {
+    const {
+      highWaterMark,
+      defaultEncoding,
+    } = options;
+    super({
+      highWaterMark,
+      defaultEncoding,
+      allowHalfOpen: true,
+      decodeStrings: true,
+      emitClose: true,
+      autoDestroy: false,
+      captureRejections: true,
+    });
+    this.#highWaterMark = highWaterMark;
+    this.#defaultEncoding = defaultEncoding;
+    this.#session = session;
+    this.#push_id = push_id;
+    this._readableState.readingMore = true;
+    this.on('pause', streamOnPause);
+
+    // See src/node_quic_stream.h for an explanation
+    // of the initial states for unidirectional streams.
+    if (this.unidirectional) {
+      if (session instanceof QuicServerSession) {
+        if (this.serverInitiated) {
+          // Close the readable side
+          this.push(null);
+          this.read();
+        } else {
+          // Close the writable side
+          this.end();
+        }
+      } else if (this.serverInitiated) {
+        // Close the writable side
+        this.end();
+      } else {
+        this.push(null);
+        this.read();
+      }
+    }
```

Comment:
If you know that the stream is not going to be `writable` or `readable` during construction it might appropriate to send `writable: false` or `readable: false` to the Duplex constructor, *unless* e.g. `_final` has some logic that needs to run. That would more closely match the behavior of `net`.

Reply:
Will address in a subsequent PR